### PR TITLE
Resources

### PR DESCRIPTION
**Pull request by TheGrimsey**
- Add player tree (kd-tree containing just player entities)
- Give players GameEntityFlagPlayer flag component and use that to update players in UpdateEntitySystem
- Give players resource,attribute,ratings components
- Temporary creature positon sync.

### CODE REVIEW

File: src/ECS/Components/Singletons/MapSingleton.h [hunk 1]
```diff
@@ -36,8 +36,10 @@ struct MapSingleton
 
 	Terrain::Map& GetCurrentMap() { return _currentMap; }
 	Tree2D& GetPlayerTree() { return _playerTree; };
+	Tree2D& GetEntityTree() { return _playerTree; };
```
Comment (NixAJ) [inline src/ECS/Components/Singletons/MapSingleton.h:39]: Looks like you made a typo here, I'm assuming GetEntityTree should return "_entityTree" and not "_playerTree"

File: src/ECS/Systems/SpawnPlayerSystem.cpp [hunk 2]
```diff
@@ -42,12 +44,25 @@ void SpawnPlayerSystem::Update(entt::registry& registry)
             Transform& transform = registry.emplace<Transform>(entityID);
             GameEntity& gameEntity = registry.emplace<GameEntity>(entityID, GameEntity::Type::Player, 29344);
             registry.emplace<TransformIsDirty>(entityID);
+            registry.emplace<GameEntityPlayerFlag>(entityID);
+
+            EntityResources& resources = registry.emplace<EntityResources>(entityID);
+            resources.resources[static_cast<u8>(EntityResourceType::HEALTH)] = 100.f;
```
Comment (NixAJ) [inline src/ECS/Systems/SpawnPlayerSystem.cpp:50]: Would it make sense given the usage here to rename the members of EntityResources to "current", "max"?

File: src/ECS/Systems/UpdateEntityPositionSystem.cpp [hunk 3]
```diff
@@ -73,21 +71,15 @@ void UpdateEntityPositionSystem::Update(entt::registry& registry)
...
 
-        // Send Movement Updates to seenEntities
+        // Send our Movement Updates to other players.
         if (seenEntities.size() > 0 && registry.all_of<TransformIsDirty>(entity))
         {
             std::shared_ptr<Bytebuffer> packetBuffer = nullptr;
             if (PacketWriter::SMSG_UPDATE_ENTITY(packetBuffer, entity, transform))
             {
-                for (u32 i = 0; i < seenEntities.size(); i++)
+                for (entt::entity seenEntity : seenEntities)
```
Comment (NixAJ) [inline src/ECS/Systems/UpdateEntityPositionSystem.cpp:80]: I personally prefer if we could move this back to a standard "for (u32 i..." and grab the entity by indexing into the vector.

File: src/EngineLoop.cpp [hunk 4]
```diff
@@ -213,13 +214,21 @@ void EngineLoop::SetupUpdateFramework()
         SpawnPlayerSystem::Update(registry);
     });
 
+    // CreatePlayerTreeSystem
+    tf::Task creatureMovementSystemTask = framework.emplace([&registry]()
+        {
```
Comment (NixAJ) [inline src/EngineLoop.cpp:219]: Please fix indentation here

**Review by NixAJ — CHANGES_REQUESTED** [review]

File: src/ECS/Systems/UpdateEntityPositionSystem.cpp [hunk 5]
```diff
@@ -77,12 +77,12 @@ void UpdateEntityPositionSystem::Update(entt::registry& registry)
             std::shared_ptr<Bytebuffer> packetBuffer = nullptr;
             if (PacketWriter::SMSG_UPDATE_ENTITY(packetBuffer, entity, transform))
             {
-                for (entt::entity seenEntity : seenEntities)
+                for (u32 i = 0; i < seenEntities.size(); i++)
                 {
-                    if (!registry.valid(seenEntity) || !registry.all_of<GameEntityPlayerFlag>(seenEntity))
```
Comment (NixAJ) [inline src/ECS/Systems/UpdateEntityPositionSystem.cpp]: Define a variable at the top of the for loop and index once, and then reuse in the rest of the loop.

**Review by NixAJ — CHANGES_REQUESTED** [review]

**Review by Pursche — APPROVED** [review]
This looks good. There is a slight concern over having to do "resources.current[static_cast<u8>(EntityResourceType::HEALTH)] = 100.f;", but we will look into a better way of doing this soon.

**Pull request merged**